nifpga/session.py

```diff
@@ -455,129 +445,162 @@ def __init__(self,
+        returns a tuple with a bool for the overflow status, and a Decimal.
 
         Returns:
-            data (DataType.value): The data in the register in a python list.
+            decimal.Decimal: Decimal representation of the fixed point number.
+        Returns:
+            (bool, decimal.Decimal): Boolean for overflow, Decimal
+                                     representation of the fixed point number.
         """
-        data = super(_FxpRegister, self).read()
-        data = self._combine_array_of_u32_into_one_value(data)
-        return self._convert_from_read_value_to_decimal(data)
+        buf = self._ctype_type()
+        self._read_func(self._session, self._resource, buf, self._transfer_len)
+        read_array = [elem for elem in buf]
+        fxp_integer = self._combine_array_of_u32_into_one_value(read_array)
+        return self._convert_from_read_value_to_decimal(fxp_integer)
 
     def _combine_array_of_u32_into_one_value(self, data):
+        """ This method is a helper to convert the array read from hardware
+        and return a single value removing any excessive bits. Whenever
+        the array is longer than 1 element, the data is left justified, we need
+        to shift the combined data to the right before doing the conversion.
+        For example, if the register had a word length of 54, the 54 MSB would
+        be the fixed point bits. The 10 LSB of the combinedData must be shifted
+        off in order to not mess up further calculations.
+
+        """
         combinedData = 0
-        for index in range(0, len(self)):
-            combinedData = (combinedData << 32 * index) + data[index]
-        if len(self) > 1:
+        for index in range(0, self._transfer_len):
+            combinedData = (combinedData << 32) + data[index]
+        if self._transfer_len > 1:
             overflow_bit = 1 if self._overflow_enabled else 0
             total_num_bits = self._word_length + overflow_bit
-            combinedData = combinedData >> (32 * len(self) - total_num_bits)
+            combinedData = combinedData >> (32 * self._transfer_len - total_num_bits)
         return combinedData
 
     def _convert_from_read_value_to_decimal(self, data):
-        """ This function converts read value from a RIO device and sets
-        the overflow bit if enabled and returns a Decimal representation of
-        the fixed point. """
+        """ This method converts value from hardware and returns the respective
+        decimal value or a tuple with the overflow status and the decimal
+        value.
+        """
+        overflow = None
         if self._overflow_enabled:
-            self.overflow = self._get_overflow_value(data)
+            overflow = self._get_overflow_value(data)
             data = self._remove_overflow_bit(data)
+
         if self._signed:
             data = self._integer_twos_comp(data)
-        return Decimal(data * self._delta)
+        decimal_value = Decimal(data * self._delta)
```

Comment:
I was expecting that _delta is always a Decimal, and that (data * self._delta) would therefore return a Decimal. Is there a cost to using that value to construct another Decimal instance?

Reply:
It looks like I misunderstand the syntax. I was expecting Decimal(<Decimal expression>) would return a new Decimal object, but that doesn't seem to be the case.